compaction/compaction_manager.cc

```diff
@@ -756,15 +758,24 @@ future<> compaction_manager::postponed_compactions_reevaluation() {
             _postponed.clear();
             co_return;
         }
-        auto postponed = std::move(_postponed);
+        // A task_state being reevaluated can re-insert itself into postponed list, which is the reason
+        // for moving the list to be processed into a local.
+        auto postponed = std::exchange(_postponed, {});
         try {
-            for (auto& t : postponed) {
+            for (auto it = postponed.begin(); it != postponed.end();) {
+                compaction::table_state* t = *it;
+                it = postponed.erase(it);
+                // skip reevaluation of a table_state that became invalid post its removal
+                if (!_compaction_state.contains(t)) {
+                    continue;
+                }
                 auto s = t->schema();
                 cmlog.debug("resubmitting postponed compaction for table {}.{} [{}]", s->ks_name(), s->cf_name(), fmt::ptr(t));
                 submit(*t);
+                co_await coroutine::maybe_yield();
             }
         } catch (...) {
-            _postponed = std::move(postponed);
+            _postponed.insert(postponed.begin(), postponed.end());
```

Comment:
I think this ends up being a re-implementation of a queue. Submitters append to the queue and wake the loop. The loop dequeues items and executes them. There's no need for the temporary postponed variable if it's really a queue.

We can follow up with converting it to a real queue (say circular_buffer)